nexus/src/db/datastore.rs

```diff
@@ -4323,14 +4323,34 @@ impl DataStore {
             .values(access_token)
             .returning(DeviceAccessToken::as_returning());
 
+        #[derive(Debug)]
+        enum TokenGrantError {
+            ConcurrentRequests,
+        }
+        type TxnError = TransactionError<TokenGrantError>;
+
         self.pool_authorized(opctx)
             .await?
             .transaction(move |conn| {
-                delete_request.execute(conn)?;
-                Ok(insert_token.get_result(conn)?)
+                if delete_request.execute(conn)? == 1 {
+                    Ok(insert_token.get_result(conn)?)
+                } else {
+                    Err(TxnError::CustomError(
+                        TokenGrantError::ConcurrentRequests,
```

Comment:
This is okay.  I would probably have this say:

0 rows back => 404 -- there's no reason to expose this as "concurrent update" -- it just doesn't exist now
more than 1 row back => 500

If you want to do that, `vpc_update_firewall_rules()` might be an example to follow?  You basically need to return something from the transaction that's specific enough to match on it in the `map_err()` below to decide whether to use `Error::internal_error` or `authz_resource.not_found()`.

Reply:
Right, sorry. Trying again in bf53d38f.

Reply:
Looks great!